src/execution_plan/optimizations/apply_join.c

```diff
@@ -0,0 +1,162 @@
+            Vector_Get(root->modifies, i, &alias);            
+            raxRemove(aliases, (unsigned char*)alias, strlen(alias), NULL);
+        }
+    }
+
+    if(raxSize(aliases) == 0) return true;
+
+    for(int i = 0; i < root->childCount; i++) {
+        if(_stream_resolves_entities(root->children[i], aliases)) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+/* filter is composed of two expressions:
+ * left and right hand side
+ * to apply a join operation each expression
+ * must be entirely resolved by one of the branches
+ * either left or right of the cartesian product operation
+ * _relate_exp_to_stream will try to associate each expression
+ * lhs and rhs with the appropriate branch. */
+static void _relate_exp_to_stream(const OpBase *cp, const FT_FilterNode *f, AR_ExpNode **lhs, AR_ExpNode **rhs) {
+    *lhs = NULL;
+    *rhs = NULL;
+    bool expression_resolved = true;
+
+    AR_ExpNode *lhs_exp = f->pred.lhs;
+    AR_ExpNode *rhs_exp = f->pred.rhs;
+
+    assert(cp->childCount == 2);
+    OpBase *left_child = cp->children[0];
+    OpBase *right_child = cp->children[1];
+
+    rax *aliases = raxNew();
+
+    /* Make sure LHS and RHS expressions can be resolved.
+     * Left expression - Left branch
+     * Right expression - Right branch */
+    AR_EXP_CollectAliases(lhs_exp, aliases);    
+    if(_stream_resolves_entities(left_child, aliases)) {
+        // aliases is now empty.
+        AR_EXP_CollectAliases(rhs_exp, aliases);        
+        if(_stream_resolves_entities(right_child, aliases)) {            
+            *lhs = lhs_exp;
+            *rhs = rhs_exp;
+            raxFree(aliases);
+            return;
+        }
+    }
+
+    /* Left expression - Right branch
+     * Right expression - Left branch */
+    *lhs = NULL;
+    *rhs = NULL;
+
+    raxFree(aliases);
+    aliases = raxNew();
+
+    AR_EXP_CollectAliases(lhs_exp, aliases);
+    if(_stream_resolves_entities(right_child, aliases)) {
+        // aliases is now empty.
+        AR_EXP_CollectAliases(rhs_exp, aliases);
+        if(_stream_resolves_entities(left_child, aliases)) {
+            *lhs = rhs_exp;
+            *rhs = lhs_exp;
+            raxFree(aliases);
+            return;
+        }
+    }
+
+    // Couldn't completely resolve either lhs_exp or rhs_exp.
+    *lhs = NULL;
+    *rhs = NULL;
+    raxFree(aliases);
+}
+
+/* Try to replace cartesian product with a value hash join operation */
+void applyJoin(ExecutionPlan *plan) {
+    OpBase **cps = ExecutionPlan_LocateOps(plan->root, OPType_CARTESIAN_PRODUCT);
+    int cp_count = array_len(cps);
+
+    for(int i = 0; i < cp_count; i++) {
+        OpBase *cp = cps[i];
+        /* TODO: change the way our cartesian product
+         * today we alow for cartesian product with > 2 child ops
+         * I think it might be better to create a chain of cartesian products
+         * where each pulls from exactly 2 streams
+         * consider: MATCH a,b,c WHERE a.v = b.v. */
```

Comment:
I agree with this idea, but it may be better to build maximal Cartesian Product ops like we do now, then decompose them in this area. (So that we can determine which sets of 2 children, if any, can be replaced by a join op.)